test/functional/testlib.bash

```diff
@@ -0,0 +1,502 @@
+create_tar() {
+
+	local item=$1
+	local path
+	local item_name
+	validate_item "$item"
+
+	path=$(dirname "$(realpath "$item")")
+	item_name=$(basename "$item")
+	# if the item is a directory exclude its content when taring
+	if [ -d "$item" ]; then
+		sudo tar -C "$path" -cf "$path"/"$item_name".tar --exclude="$item_name"/* "$item_name"
+	else
+		sudo tar -C "$path" -cf "$path"/"$item_name".tar "$item_name"
+	fi
+
+}
+
+# Creates an empty manifest in the specified path
+# Parameters:
+# - PATH: the path where the manifest will be created
+# - BUNDLE_NAME: the name of the bundle which this manifest will be for
+create_manifest() {
+
+	local path=$1
+	local name=$2
+	local version
+	validate_path "$path"
+	validate_param "$name"
+
+	version=$(basename "$path")	
+	echo -e "MANIFEST\\t3" | sudo tee "$path"/Manifest."$name" > /dev/null
+	echo -e "version:\\t$version" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "previous:\\t0" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "filecount:\\t0" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "timestamp:\\t$(date +"%s")" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo -e "contentsize:\\t0" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo "" | sudo tee -a "$path"/Manifest."$name" > /dev/null
+	echo "$path/Manifest.$name"
+	
+}
+
+# Adds the specified item to an existing bundle manifest
+# Parameters:
+# - MANIFEST: the relative path to the manifest file
+# - ITEM: the relative path to the item (file, directory, symlink) to be added
+# - ITEM_PATH: the absolute path of the item in the target system when installed
+add_to_manifest() { 
+
+	local manifest=$1
+	local item=$2
+	local item_path=$3
+	local item_type
+	local item_size
+	local name
+	local version
+	local filecount
+	local contentsize
+	local linked_file
+	validate_item "$manifest"
+	validate_item "$item"
+	validate_param "$item_path"
+	
+	item_size=$(du -b "$item" | cut -f1)
+	name=$(basename "$item")
+	version=$(basename "$(dirname "$manifest")")
+	# add to filecount
+	filecount=$(sudo cat "$manifest" | grep filecount | awk '{ print $2 }')
+	filecount=$((filecount + 1))
+	sudo sed -i "s/filecount:.*/filecount:\\t$filecount/" "$manifest"
+	# add to contentsize 
+	contentsize=$(sudo cat "$manifest" | grep contentsize | awk '{ print $2 }')
+	contentsize=$((contentsize + item_size))
+	# get the item type
+	if [ "$(basename "$manifest")" = Manifest.MoM ]; then
+		item_type=M
+		# MoM has a contentsize of 0, so don't increase this for MoM
+		contentsize=0
+		# files, directories and links are stored already hashed, but since
+		# manifests are not stored hashed, we need to calculate the hash
+		# of the manifest before adding it to the MoM
+		name=$(sudo "$SWUPD" hashdump "$item" 2> /dev/null)
+	elif [ -L "$item" ]; then
+		item_type=L
+		# when adding a link to a bundle, we need to make sure we add
+		# its associated file too
```

Comment:
We probably want to add the associated file, but I could see that we might want dangling links ....

Reply:
Can you elaborate a little more please? I don't know what are dangling links or how to create/use those

Reply:
A dangling symbolic link is a symbolic link that points to a file that doesn't exist.

You can create one with `ln -s /does_not_exist dangle` - after which `dandle` exists in the current directory but if you try and open it the open will fail with EEXIST as `/does_not_exist` is assumed not to exist.